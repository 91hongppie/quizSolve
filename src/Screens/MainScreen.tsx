import {Ref, useRef, useState} from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  TextInput,
  KeyboardAvoidingView,
  StyleSheet,
  Alert,
} from 'react-native';
import api from '../apis/baseClient';
import LocalStorage from '../utils/LocalStorage';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../App';

type MainScreenProps = NativeStackScreenProps<RootStackParamList, 'Main'>;

const MainScreen = ({navigation}: MainScreenProps) => {
  const [nickname, setNickname] = useState('홍길동');
  const [amount, setAmount] = useState<number>();
  const amountInput = useRef<TextInput>(null);

  const handleAmount = (next: string) => {
    const parsedAmount = parseInt(next);
    if (isNaN(parsedAmount)) {
      setAmount(undefined);
    } else if (parsedAmount > 50) {
      setAmount(50);
    } else {
      setAmount(parsedAmount);
    }
  };

  const moveCursor = () => {
    if (!amountInput.current) {
      return;
    }
    amountInput.current.focus();
  };

  const handleNickname = (e: string) => {
    setNickname(e);
  };

  const handleQuiz = async () => {
    try {
      const questionAmount = amount || 10;
      const prevQuizs = await LocalStorage.getItem('quizs');

      let quizs;
      if (prevQuizs) {
        quizs = prevQuizs;
      } else {
        const res = await api.Quiz.getQuiz({amount: questionAmount});
        quizs = res['results'];
      }
      await LocalStorage.setItem('quizs', quizs);
      await LocalStorage.setItem('nickname', nickname);
      navigation.navigate('Quiz', {quizNum: 0});
    } catch (e) {
      console.log(e);
      Alert.alert('잠시 후 다시 시도해주세요.');
    }
  };

  return (
    <KeyboardAvoidingView style={styles.container}>
      <View style={{flex: 1}}></View>
      <View style={styles.descContainer}>
        <Text style={{fontSize: 23}}>안녕하세요 </Text>
        <TextInput
          style={styles.textInput}
          testID={'nicknameInput'}
          placeholder="홍길동"
          maxLength={5}
          multiline={false}
          textAlign="center"
          onChangeText={handleNickname}
          onSubmitEditing={moveCursor}
        />
        <Text style={{fontSize: 23}}>님</Text>
      </View>
      <View style={{flex: 0.1}}></View>
      <View style={styles.descContainer}>
        <Text style={{fontSize: 23}}>문제</Text>
        <TextInput
          testID={'quizNumInput'}
          style={styles.textInput}
          ref={amountInput}
          placeholder="10"
          value={amount !== undefined ? `${amount}` : ''}
          multiline={false}
          keyboardType="numeric"
          textAlign="center"
          onChangeText={handleAmount}
        />
        <Text style={{fontSize: 23}}>개 풀어보실래요?</Text>
      </View>
      <View style={{flex: 1}}></View>
      <TouchableOpacity
        style={styles.button}
        onPress={handleQuiz}
        testID="quizSolveButton">
        <Text style={{fontSize: 25, color: 'white'}}>문제 풀기</Text>
      </TouchableOpacity>
      <View style={{flex: 0.5}}></View>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container: {height: '100%'},
  descContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  textInput: {
    width: 100,
    fontSize: 23,
    borderBottomWidth: 1,
    borderBottomColor: 'black',
  },
  button: {
    alignItems: 'center',
    backgroundColor: 'green',
    width: '90%',
    marginLeft: 'auto',
    marginRight: 'auto',
    height: 60,
    justifyContent: 'center',
    borderRadius: 10,
  },
});

export default MainScreen;
