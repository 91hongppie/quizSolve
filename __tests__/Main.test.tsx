import 'react-native';
import React, {ReactElement} from 'react';

// Note: import explicitly to use the types shipped with jest.
import {it, describe} from '@jest/globals';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import MainScreen from '../src/Screens/MainScreen';
import {render} from '@testing-library/react-native';

let props;
let component: ReactElement;

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

function getTempComponent(props: any) {
  return <MainScreen {...props} />;
}
describe('Main', () => {
  props = {};
  component = getTempComponent(props);
  it('renders without crashing', () => {
    const renderResult = render(component);
    const renderedJson = renderResult.toJSON();
    expect(renderedJson).toMatchSnapshot();
    expect(renderedJson).toBeTruthy();
  });

  it('닉네임 입력창 존재 확인', () => {
    const renderResult = render(component);

    expect(renderResult.getByTestId('nicknameInput')).toBeDefined();
  });
  it('문항 수 입력창 존재 확인', () => {
    const renderResult = render(component);

    expect(renderResult.getByTestId('quizNumInput')).toBeDefined();
  });
  it('문제풀기 버튼 존재 확인', () => {
    const renderResult = render(component);

    expect(renderResult.getByTestId('quizSolveButton')).toBeDefined();
  });
});
