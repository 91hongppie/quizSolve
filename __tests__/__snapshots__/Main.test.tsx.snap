// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders without crashing 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "height": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 23,
        }
      }
    >
      안녕하세요 
    </Text>
    <TextInput
      maxLength={5}
      multiline={false}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="홍길동"
      style={
        {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "fontSize": 23,
          "width": 100,
        }
      }
      testID="nicknameInput"
      textAlign="center"
    />
    <Text
      style={
        {
          "fontSize": 23,
        }
      }
    >
      님
    </Text>
  </View>
  <View
    style={
      {
        "flex": 0.1,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 23,
        }
      }
    >
      문제
    </Text>
    <TextInput
      keyboardType="numeric"
      multiline={false}
      onChangeText={[Function]}
      placeholder="10"
      style={
        {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "fontSize": 23,
          "width": 100,
        }
      }
      testID="quizNumInput"
      textAlign="center"
      value=""
    />
    <Text
      style={
        {
          "fontSize": 23,
        }
      }
    >
      개 풀어보실래요?
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "green",
        "borderRadius": 10,
        "height": 60,
        "justifyContent": "center",
        "marginLeft": "auto",
        "marginRight": "auto",
        "opacity": 1,
        "width": "90%",
      }
    }
    testID="quizSolveButton"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 25,
        }
      }
    >
      문제 풀기
    </Text>
  </View>
  <View
    style={
      {
        "flex": 0.5,
      }
    }
  />
</View>
`;
